forks/cli-artifacts/Dockerfile.centos9

```diff
@@ -1,22 +1,89 @@
 # This Dockerfile builds an image containing the Mac and Windows version of oc
 # layered on top of the Linux cli image.
 FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.18-openshift-4.11 AS builder
-WORKDIR /go/src/github.com/openshift/oc
+
+ARG PKG_NAME=openshift-clients
+ARG INSTALL_PKGS="krb5-devel"
+ARG SPEC_FILENAME="oc.spec"
+ARG SPEC_FILEPATH="./"
+ARG PKG_GO_PATH=/go/src/github.com/openshift/oc
+ARG BINS="oc"
+ARG ARCHITECTURES="x86_64 aarch64 ppc64le s390x"
+
+WORKDIR ${PKG_GO_PATH}
 COPY . .
-RUN yum install -y krb5-devel && yum clean all
-RUN GOPATH=$HOME/go GOFLAGS="" go install github.com/josephspurrier/goversioninfo/cmd/goversioninfo@latest && mkdir $HOME/bin && cp -rvf $HOME/go/bin/goversioninfo /usr/local/bin
-RUN make cross-build --warn-undefined-variables
+RUN yum install -y --setopt=skip_missing_names_on_install=False \
+            rpm-build \
+            createrepo \
+            bsdtar \
+            rsync ${INSTALL_PKGS}
 
+# Sketching sol. 1
```

Comment:
Hello, I'm trying to generalize the `*-artifacts` images to have them buildable from a single "dockerfile".

Based on what we disccused, it would be nice to have a single dockerfile to build multiarch binaries and RPMs for both oc and hyperkube.

Doing it for the RPMs is ok. However, binaries are built differently:

1. hypershift: KUBE_BUILD_PLATFORMS="...." make WHAT...
2. oc: make (cross-)build 

The output in the build layer would be different, but manageable:
1. hypershift: _output/local/bin/linux/%arch%/{bins}
2. oc: _output/local/bin/%os%_%arch%

The runtime stage will also need some trick to have the final multi-arch binaries in the place where their consumers look for them. See the [cli-artifacts dockerfile](https://github.com/openshift/oc/blob/af05172ab6dbb11dd3309268cf486fb3e482e7ce/images/cli-artifacts/Dockerfile.rhel) to have an idea.

**Solution1**: Generlize by using an arg that provides the whole command to run the build. 
In this case, we'd build twice for each architecture, rpm + single binary. It seems not elegant and slow to me. Isn't it?

**Solution2**: We'd have to extract each rpm for each architecture, get the executables and place them accordingly in their desired final dir for the payload's image.
The build will be faster as there is no double build, and no "eval-like" code has to be provided in a build argument. 

The drawback is for `oc`. When the arch is x86_64, the rpmbuild command creates two RPMs, one having the oc binary for win and x86_64 mac.
Then, we'd loose the darwin/arm64 binary. <- something should change here eventually, but we cannot make this solution the upstream one currently.

Solution3: Put the script to generalize the RPM building process in an auxiliary image (the builder imo) and continue maintaining two different and smaller Dockerfiles, hyperkube-artifacts and cli-artifacts, to join together in the artifacts one.

We still use the double build process, but since the Dockerfiles need to be kept in two repos, we will have a more homogeneous building process. The double build could be slightly improved in performance maybe.

Do you have any hint?

Reply:
Great overview, thanks for providing that context! Sounds to me like Solution3 would be the sanest then